Category Listing Screen - UI Implementation

AjeshRPai opened the pull request:
Parent #15550

# Description
This PR adds the Implementation of the UI for the Category Listing Screen.  As part of the task, this PR will have the code the following changes in code 

## Additions 

- Adds a new layout XML file `site_settings_categories_row.xml`, Which represents the design for the RecyclerView Row 
- Adds Styles to `styles.xml` for the Layout and TextView in the `site_settings_categories_row` 
- Adds supporting Kotlin classes for RecyclerView  
    - Adapter - `SiteSettingsCategoriesAdapter` 
    - ViewHolder - `SiteSettingsCategoriesViewHolder`
    - Diffcallback  - `SiteSettingsCategoriesDiffCallback`

## Updates  

#### `CategoriesListFragment`
-  Adds function to initialize the Recyclerview with ``SiteSettingsCategoriesAdapter`` and item Decoration 
-  Adds function to show the states according to the Observing Livedata 
-  Adds function to pass to `SiteSettingsCategoriesAdapter` which is invoked on clicking at List Row 

#### `CategoriesListViewModel`
- Adds function, Which will be invoked from `CategoriesListFragment` on clicking at Categories List Row 
- Updates `UiState` class to hold the information about the visibility of views 

#### `site_settings_categories_list_fragment.xml`
- Adds `ProgressBar` to be shown in the loading state


## UI
Screenshots from the Testing Device 

There are three mutually exclusive states for the Category Listing Screen. They are

1. Loading 
2. Content 
3. Error 
    3.1. No Network Error 
    3.2. Generic Error 


| Loading(1)|Content(2)|No Network error(3.1)|Generic Error (3.2)|
| :---        |    :----:   |          ---: |          ---: |
| ![Screenshot_20211116-190107](https://user-images.githubusercontent.com/17463767/142137306-92ba38f7-2bc2-4541-8cb2-f61879aab215.jpg)      | ![Screenshot_20211116-184130](https://user-images.githubusercontent.com/17463767/142137222-e6907595-3856-4544-9c64-728066ba7f8e.jpg)       | ![Screenshot_20211116-190830](https://user-images.githubusercontent.com/17463767/142137401-003503cb-d7a0-43dc-b681-100c2272eb2a.jpg)   | ![Screenshot_20211116-190104](https://user-images.githubusercontent.com/17463767/142137547-b2e7325e-a702-469c-8492-4a51c508ca99.jpg) |


## Yet to be added 
- The navigation on clicking at the Category in the List 
- Show the FAB to create a Category 

These will be taken up in the future in a separate PR

# Testing Instructions

## Prerequisite:
- Go to My Site -> Me  -> App Settings -> Test Feature Configuration
- Enable ManageCategoriesFeatureConfig and restart the app
- Relaunch the app.

## To test:

Test that loading screen is displayed (1) and Content(2) is displayed
- Fresh install the app and logon to the wp.com account 
- Go to my My Site -> Site Settings -> Categories 
- Note that the Screen Displays Loading state first followed by Categories Content 
PS: If no categories are present for the Account, then a default Category - **UnCategorized** exists for an account 

----

These states can be tested by using CategoriesListViewModel tests:

### No Network error(3.1)
#### Categories Screen being accessed for the First Time - 
If the internet is not available, The Site settings page disables all the editing. ie the user won't be able to click the Categories row and navigate to Categories Listing Screen. 
For instance, the Internet is available and the User navigates to the Categories Listing screen and if the Internet is disconnected just before the API call, this Scenario would occur . 

This state can be tested using the Unit test - 
`given no items in db with no internet available, when vm starts, then no connection error is displayed`

PS: Note In this case the categories were able to fetch successfully at least once, the items would be displayed from DB and no network error won't be displayed
This state can be tested using the Unit test - 
`when vm starts, then list of items from db is displayed`

### Generic Error Failed (3.2)

This state can be tested by using`CategoriesListViewModel` tests:
`given no items in db and api error occurs, when vm starts, then error is displayed`

## Regression Notes

1. Potential unintended areas of impact
None


2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual tests

3. What automated tests I added (or what prevented me from doing so)
None


PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

AjeshRPai committed:
Merge branch 'wordpress-mobile:develop' into issue/5382-categories-list-ui-part-i

AjeshRPai committed:
Add: Loading Indicator

Adds a Progress bar to the Layout to show the Loading State

AjeshRPai committed:
Refactor: Implements processing of UI state in CategoriesListFragment

Refactors the Logic of setting the visibility of views corresponding to the UI state
Adds the Logic of Initializing the Recyclerview and Empty State view

AjeshRPai committed:
Merge branch 'develop' into issue/5382-categories-list-ui-part-i

# Conflicts:
#	WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListFragment.kt

AjeshRPai committed:
Removes : Unused Import

AjeshRPai committed:
Adds : Row XML for Categories Listing

Adds the file site_settings_categories_row.xml which is used in listing the Categories

AjeshRPai committed:
Adds: Recycler ViewHolder for Categories Listing UI

Adds SiteSettingsCategoriesViewHolder which will be used in the Recycler Adapter for showing the Categories in the CategoriesListFragment

AjeshRPai committed:
Adds: Recycler Adapter for Showing Categories

Adds SiteSettingsCategoriesAdapter for showing Categories. Uses SiteSettingsCategoriesViewHolder.
Adds SiteSettingsCategoriesDiffUtils to handle the difference of new list with previous list

AjeshRPai committed:
Updates: CategoriesListFragment with RecyclerView Adapter

Updates CategoriesListFragment to use SiteSettingsCategoriesAdapter to show the List of categories .

AjeshRPai committed:
Updates : RecyclerViewAdapter with ListAdapter

- Updates: SiteSettingsCategoriesAdapter was using RecyclerView.Adapter as base class, In which we had to implement the logic of calling the DiffUtil and dispatching the result. And it used to happen on the Main thread. This is Changed to ListAdapter as that would take care of the DiffUtil in the background.
- Moves: SiteSettingsCategoriesDiffCallback to a Separate class
- Updates: SiteSettingsCategoriesViewHolder to handle the changes on Updating list

AjeshRPai committed:
Updates: Categories Row to support RTL rendering

AjeshRPai committed:
Updates : SiteSettingsCategoriesAdapter

Removes: Function getItemId as it is no longer needed since the DiffUtil calculates the difference between the new and old list. Stable id's are redundant. Ref - https://medium.com/android-news/anatomy-of-recyclerview-part-1-a-search-for-a-viewholder-continued-d81c631a2b91
Updates: SiteSettingsCategoriesDiffCallback for correct logic in areContentsTheSame. What was i thinking ? :)

AjeshRPai committed:
Adds: Click Listener Function to the Categories Listing Row

Adds: Click Listener to the SiteSettingsCategoriesViewHolder which will be passed from the Fragment to the Adapter and then to the ViewHolder.
Adds: Function in CategoriesListFragment and CategoriesListViewModel which will be invoked on Clicking at the Category Row

AjeshRPai committed:
Fixes: Formatting and Lint Issues

AjeshRPai committed:
Adds: New Style for CategoryName and Row

AjeshRPai committed:
Fixes: Error Image not being Displayed

AjeshRPai committed:
Adds: Divider height to RecyclerView

AjeshRPai committed:
Removes: Redundant Layout Manager from Fragment

Removes: Linear layout manager from fragment as that is being set in the Layout XML

AjeshRPai committed:
Removes: Suppress warning

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListFragment.kt:
```diff
@@ -45,10 +54,78 @@ class CategoriesListFragment : Fragment(R.layout.site_settings_categories_list_f
+        fabButton.updateVisibility(false)
+        progressBar.updateVisibility(false)
+        categoriesRecyclerView.updateVisibility(false)
+    }
```
Minor 🔍 : We try not to add view visibility logic in the fragment but get it from the `ViewModel` (see [this](https://github.com/wordpress-mobile/WordPress-Android/blob/61951f769fb70534c3583764a7795b6bf906039f/WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanViewModel.kt#L350-L355) as an example).

AjeshRPai replied:
@ashiagr Great suggestion 🙌 . I had seen this logic but thought it would be an overkill. I see the point now, by updating the `UiSstate` to hold that logic, the `Fragment` looks better.  Thank you for this suggestion. 👍 
Done in https://github.com/wordpress-mobile/WordPress-Android/pull/15580/commits/e620061108e8a176167ec2046883445641410d0e

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/CategoriesListFragment.kt:
```diff
@@ -45,10 +54,78 @@ class CategoriesListFragment : Fragment(R.layout.site_settings_categories_list_f
+        categoriesRecyclerView.setEmptyView(actionableEmptyView)
+        actionableEmptyView.updateVisibility(false)
+    }
+
```
Minor 🔍 : We can move these `init` functions to the top along with other initialization functions.

AjeshRPai replied:
Yes, You are right. That would look cleaner and better . Done in https://github.com/wordpress-mobile/WordPress-Android/pull/15580/commits/c3d988502a6366abfead711fcaa4dd056fd18d2d . 😄

ashiagr commented on WordPress/src/main/res/layout/site_settings_categories_list_fragment.xml:
```diff
@@ -17,8 +17,19 @@
+        android:visibility="gone"
         app:layoutManager="androidx.recyclerview.widget.LinearLayoutManager"
         app:layout_constraintBottom_toBottomOf="parent"
+        android:dividerHeight="@dimen/divider_size"
```
Minor 🔍 : `android:dividerHeight="@dimen/divider_size"` can go above `android:visibility="gone"`. You can verify it by reformatting the xml but rearranging the elements to their screen positions.

AjeshRPai replied:
Done in https://github.com/wordpress-mobile/WordPress-Android/pull/15580/commits/2abec5c8dacbaa5027d390816724ffb2e5ed511e

ashiagr commented on WordPress/src/main/res/layout/site_settings_categories_row.xml:
```diff
@@ -0,0 +1,21 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:text="Category Name" />
```
Super Minor 🔍 : We can reuse a preexisting string from the `strings.xml` here `@string/category_name`.

AjeshRPai replied:
@ashiagr  Nice one, Changed in  f7b54da

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/SiteSettingsCategoriesAdapter.kt:
```diff
@@ -0,0 +1,34 @@
+class SiteSettingsCategoriesAdapter(
+    private val uiHelpers: UiHelpers,
+    private val onClickListener: (CategoryNode) -> Unit
+) : ListAdapter<CategoryNode, SiteSettingsCategoriesViewHolder>(SiteSettingsCategoriesDiffCallback) {
```
Nice touch on using a [ListAdapter](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter) here. 👍

ashiagr commented on WordPress/src/main/java/org/wordpress/android/ui/prefs/categories/SiteSettingsCategoriesViewHolder.kt:
```diff
@@ -0,0 +1,53 @@
+    private val horizontalPadding: Int = uiHelpers.getPxOfUiDimen(
+            categoryBinding.siteSettingsCategoryText.context,
+            UIDimenRes(R.dimen.margin_extra_large)
+    )
```
Minor Suggestion: Drop explicit `Int` declaration and directly use `itemView.context`.

```suggestion
    private val verticalPadding = uiHelpers.getPxOfUiDimen(itemView.context, UIDimenRes(R.dimen.margin_large))
    private val horizontalPadding = uiHelpers.getPxOfUiDimen(itemView.context, UIDimenRes(R.dimen.margin_extra_large))
```

AjeshRPai replied:
Done in https://github.com/wordpress-mobile/WordPress-Android/pull/15580/commits/955f8bfb9c4d7344138f828edbce81231e3fb9da

ashiagr reviewed (commented):
Nice job @AjeshRPai! 

I liked the fact that you followed our existing code conventions like the `viewBinding` for the `ViewHolders`, and at the same time brought small improvements by using a more performant [ListAdapter](https://developer.android.com/reference/androidx/recyclerview/widget/ListAdapter).

I also liked that due attention was paid to the RTL rendering and how an [existing RTL bug was found in the PPN dialog categories ](https://github.com/wordpress-mobile/WordPress-Android/issues/15581)during implementation. ✨

I just left minor comments and suggestions for you to consider. 

There's also one ⚠️ that I want to highlight: the `retry` button doesn't work right now. We might want to add it to our `TODO` list so that we don't forget about it as the network error view cannot be easily seen because of a potential `Site Settings` screen bug we discussed in the DM.

AjeshRPai committed:
Updates: Variable initialization with Itemview context

Co-authored-by: ashiagr <agr.ash@gmail.com>

AjeshRPai committed:
Updates: Layout to use pre-existing string resource

AjeshRPai committed:
Fixes: Formatting of View properties

AjeshRPai committed:
Moves : Initialization functions to top

AjeshRPai committed:
Updates: UI state to hold view visibility

Updates the UI state to hold the visibility information of views there by removing the logic of visibility of views from the fragment.

AjeshRPai commented:
> There's also one ⚠️ that I want to highlight: the `retry` button doesn't work right now. We might want to add it to our `TODO` list so that we don't forget about it as the network error view cannot be easily seen because of a potential `Site Settings` screen bug we discussed in the DM.

@ashiagr 🤔 I had manually tested the no network state by hardcoding No Network error on VM start and then clicking on the `retry` button. It was working as expected when there is internet available. 

When the internet is not available, clicking the "retry" button shows no loading indicator, and since the check for showing error in the no internet condition is to check whether loading is displayed, No Network error won't be shown as well. This is a bug 😟 . 

Since this is due to the incorrect Implementation of `fetchCategoriesFromNetwork`, we can take it up in a separate PR as this will require only changes in the `CategoriesListViewModel`. Tests for this scenario will be added in `CategoriesListViewModelTest` as well.  We can add this in #15550 as a TODO. WDYT ? 🤔 

FYI Another similar scenario where this bug can occur is 
1. When No Items are Present in DB 
2. And Generic Error Occurs
3. On Clicking Retry Button 
4. If No Internet is available then it will show the Generic error state itself.  

This can also be taken care of while fixing the No Network error.

ashiagr reviewed (commented):
> When the internet is not available, clicking the "retry" button shows no loading indicator, and since the check for showing error in the no internet condition is to check whether loading is displayed, No Network error won't be shown as well. This is a bug 😟 .

> Since this is due to the incorrect Implementation of fetchCategoriesFromNetwork, we can take it up in a separate PR as this will require only changes in the CategoriesListViewModel. Tests for this scenario will be added in CategoriesListViewModelTest as well. We can add this in #15550 as a TODO. WDYT ? 🤔

Thanks for verifying it, @AjeshRPai. 👍  
And yes, we can add it to the TODO list and take it up separately. Once it is tracked in TODO, I'll approve this PR and initiate a merge to `develop`.

AjeshRPai commented:
@ashiagr  Thanks for the detailed review and for finding out the "Retry" button issue. 🙌 👍 . `retry` button Issues are added in #15550 as TODO.

AjeshRPai committed:
Adds: Suppress Warning for Detekt Issue

ashiagr commented:
👋 Hey @osullivanchris!

Pinging you in case you want to take a look at the categories listing screen with error handling (accessible from the `My Site` ->  `Site Settings` -> `Categories`). The screen lists categories without the selection checkboxes and displays error messages in the standard format.

osullivanchris commented:
Thanks for the ping @ashiagr 

Looks good so far 👍
